frontend/packages/pipelines-plugin/integration-tests/support/step-definitions/pipelines/pipelines-runs.ts

```diff
@@ -473,3 +479,108 @@ Then(
     modal.shouldBeClosed();
   },
 );
+
+Given('user is at pipeline page in developer perspective', () => {
+  perspective.switchTo(switchPerspective.Developer);
+});
+
+Given('a failed pipeline is present', () => {
+  navigateTo(devNavigationMenu.Add);
+  addPage.selectCardFromOptions(addOptions.ImportFromGit);
+  gitPage.enterGitUrl('https://github.com/sclorg/golang-ex');
+  devFilePage.verifyValidatedMessage('https://github.com/sclorg/golang-ex');
+  gitPage.selectAddPipeline();
+  gitPage.clickCreate();
+  topologyPage.verifyTopologyPage();
+});
+
+When('user goes to failed pipeline run of pipeline {string}', (pipelineName: string) => {
+  navigateTo(devNavigationMenu.Pipelines);
+  pipelinesPage.search(pipelineName);
+  cy.get(`[data-test-id^="${pipelineName}-"]`).click();
+});
+
+When('user opens pipeline run details', () => {
+  cy.get('.pf-c-breadcrumb').should('include.text', 'PipelineRun details');
+});
+
+Then('user can see status as Failure', () => {
+  cy.get(pipelineRunsPO.pipelineRunsTable.status).should('include.text', 'Failed');
+});
+
+Then('user can view failure message under Message heading', () => {
+  cy.get('.odc-pipeline-run-details__customDetails').within(() => {
+    cy.get('dl dt')
+      .contains('Message')
+      .should('be.visible');
+  });
+});
+
+Then('user can see Log snippet to get know what taskruns failed', () => {
+  cy.get('.odc-pipeline-run-details__customDetails').within(() => {
+    cy.get('dl dt')
+      .contains('Log snippet')
+      .should('be.visible');
+  });
+});
+
+Given('user has passed pipeline run', () => {
+  cy.exec(
+    `oc apply -f testData/pipelines-workspaces/sum-three-pipeline.yaml -n ${Cypress.env(
+      'NAMESPACE',
+    )}`,
+  );
+});
+
+When(
+  'user is on Pipeline Run details page of {string} pipeline run',
+  (pipelineRunsName: string) => {
+    cy.get(`[data-test-id^="${pipelineRunsName}"]`).click();
+  },
+);
+
+When('user scrolls to the Pipeline Run results section', () => {
+  cy.get('[data-test-section-heading="PipelineRun results"]').should('be.visible');
+});
+
+Then('user can see Name and Value column under Pipeline Run results', () => {
+  cy.get('[role="grid"]').should('contain.text', 'Name');
+  cy.get('[role="grid"]').should('contain.text', 'Value');
+});
+
+When(
+  'a node with an associated pipeline {string} that has failed is present',
+  (pipelineName: string) => {
+    topologyPage.componentNode(pipelineName).should('be.visible');
+  },
+);
```

Comment:
- this is a `Given` step
- shouldn't we verify the node shows a failed pipeline here?